server/src/main/java/org/opensearch/gateway/remote/model/AbstractRemoteBlobStore.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.gateway.remote.model;
+
+import static org.opensearch.gateway.remote.RemoteClusterStateUtils.PATH_DELIMITER;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Arrays;
+import java.util.concurrent.ExecutorService;
+import org.opensearch.common.blobstore.BlobPath;
+import org.opensearch.common.blobstore.stream.write.WritePriority;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.gateway.remote.RemoteClusterStateUtils;
+import org.opensearch.index.translog.transfer.BlobStoreTransferService;
+import org.opensearch.repositories.blobstore.BlobStoreRepository;
+import org.opensearch.threadpool.ThreadPool;
+
+/**
+ * Abstract class for a blob type storage
+ *
+ * @param <T> The entity which can be uploaded to / downloaded from blob store
+ * @param <U> The concrete class implementing {@link RemoteObject} which is used as a wrapper for T entity.
+ */
+public class AbstractRemoteBlobStore<T, U extends AbstractRemoteBlobObject<T>> implements RemoteObjectStore<T, U> {
+
+    private final BlobStoreTransferService transferService;
+    private final BlobStoreRepository blobStoreRepository;
+    private final String clusterName;
+    private final ExecutorService executorService;
+
+    public AbstractRemoteBlobStore(BlobStoreTransferService blobStoreTransferService, BlobStoreRepository blobStoreRepository, String clusterName,
+        ThreadPool threadPool) {
+        this.transferService = blobStoreTransferService;
+        this.blobStoreRepository = blobStoreRepository;
+        this.clusterName = clusterName;
+        this.executorService = threadPool.executor(ThreadPool.Names.GENERIC);
```

Comment:
Lets use a configurable thread pool, lets not use `GENERIC`

Reply:
I can add another parameter for `Threadpool.Names` to keep it configurable.